File: code/fig2_calibration.R
#!/usr/bin/env Rscript
# Figure 2 — calibration (A: Any-SHD reliability before/after recalibration;
# B: per-condition over-prediction). Reads CSVs from fig2_calibration.py.
# Run: python3 code/fig2_calibration.py  &&  Rscript code/fig2_calibration.R
suppressMessages({library(ggplot2); library(patchwork)})
navy<-"#1f3b5c"; orange<-"#d8862b"; grey<-"#9aa6b2"
fig <- file.path(Sys.getenv("ECHONEXT_DATA", path.expand("~/Desktop/RESEARCH")), "EchoNext-repo/results_full/figs")
repo<- file.path(Sys.getenv("ECHONEXT_DATA", path.expand("~/Desktop/RESEARCH")), "echonext-mimic-validation/results/figures")
rel<-read.csv(file.path(fig,"reliability.csv"), check.names=FALSE, encoding="UTF-8")
cil<-read.csv(file.path(fig,"cil.csv"), encoding="UTF-8"); cil<-cil[order(cil$cil),]
cil$label<-factor(cil$label, levels=cil$label); cil$grp<-ifelse(cil$is_shd==1,"Composite","Component")

pA<-ggplot(rel, aes(pred, obs, color=method, shape=method))+
  geom_abline(slope=1,intercept=0,linetype="dashed",color="grey55",linewidth=0.4)+
  geom_line(linewidth=0.8)+geom_point(size=2.1)+
  scale_color_manual(values=c("Original (slope 0.94)"=navy,"Recalibrated (out-of-fold Platt)"=orange))+
  scale_shape_manual(values=c("Original (slope 0.94)"=16,"Recalibrated (out-of-fold Platt)"=15))+
  coord_equal(xlim=c(0,1),ylim=c(0,1),expand=FALSE)+
  labs(x="Predicted probability",y="Observed frequency",title="Any-SHD calibration",color=NULL,shape=NULL)+
  theme_classic(base_size=11,base_family="Helvetica")+
  theme(legend.position=c(0.03,0.99),legend.justification=c(0,1),legend.background=element_blank(),
        legend.key.height=unit(12,"pt"),plot.title=element_text(size=11),plot.margin=margin(6,12,6,6))

pB<-ggplot(cil, aes(cil, label, fill=grp))+
  geom_col(width=0.72)+geom_vline(xintercept=0,color="black",linewidth=0.4)+
  scale_fill_manual(values=c("Composite"=navy,"Component"=grey),guide="none")+
  labs(x="Calibration-in-the-large",y=NULL,title="Per-condition over-prediction")+
  theme_classic(base_size=11,base_family="Helvetica")+
  theme(plot.title=element_text(size=11),plot.margin=margin(6,8,6,6))

p<-pA+pB+plot_layout(widths=c(1,1.15))+plot_annotation(tag_levels="A")&
  theme(plot.tag=element_text(face="bold",size=15,family="Helvetica"))
for (d in c(repo, fig)) { if(!dir.exists(d)) dir.create(d,recursive=TRUE)
  ggsave(file.path(d,"fig3_calibration.png"), p, width=10, height=4.7, dpi=300, bg="white")
  ggsave(file.path(d,"fig3_calibration.pdf"), p, width=10, height=4.7, bg="white") }
cat("saved fig3_calibration (R/ggplot2)\n")

File: code/fig2_calibration.py
#!/usr/bin/env python3
"""Export plot-ready calibration data for Figure 2 (rendered by fig2_calibration.R).
Writes reliability.csv (composite reliability curve, original + OOF-Platt) and
cil.csv (per-condition calibration-in-the-large) to results_full/figs/."""
import os, csv, numpy as np
from sklearn.calibration import calibration_curve
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression
RES=os.path.join(os.environ.get("ECHONEXT_DATA", os.path.expanduser("~/Desktop/RESEARCH")), "EchoNext-repo/results_full")
OUT=os.path.join(RES,"figs"); os.makedirs(OUT,exist_ok=True)
probs=np.load(f"{RES}/probs.npy"); paths=[p for p in open(f"{RES}/kept_paths.txt").read().split("\n") if p]
coh={r['ecg_path']:r for r in csv.DictReader(open(os.path.join(os.environ.get("ECHONEXT_DATA", os.path.expanduser("~/Desktop/RESEARCH")), "cohort_oneperpt_full.csv")))}
COLS=['lvef_lte_45','lvwt_gte_13','aortic_stenosis_modsev','aortic_regurg_modsev','mitral_regurg_modsev','tricuspid_regurg_modsev','pulm_regurg_modsev','rv_dysfunction_modsev','pericardial_modlarge','pasp_gte_45','tr_max_gte_32','shd']
NICE=['Reduced LVEF','LV hypertrophy','Aortic stenosis','Aortic regurgitation','Mitral regurgitation','Tricuspid regurgitation','Pulmonic regurgitation','RV dysfunction','Pericardial effusion','Elevated PASP','Elevated TR Vmax','Any SHD (composite)']
rows=[coh[p] for p in paths]; y=np.array([[int(r[c]) for c in COLS] for r in rows]); n=len(rows)
ys,ps=y[:,11],probs[:,11]
oof=np.zeros(n); eps=1e-6; lg=np.log(np.clip(ps,eps,1-eps)/(1-np.clip(ps,eps,1-eps)))
for tr,te in KFold(5,shuffle=True,random_state=0).split(ps):
    lr=LogisticRegression(C=1e6).fit(lg[tr].reshape(-1,1),ys[tr]); oof[te]=lr.predict_proba(lg[te].reshape(-1,1))[:,1]
x0,o0=calibration_curve(ys,ps,n_bins=12,strategy='quantile'); x1,o1=calibration_curve(ys,oof,n_bins=12,strategy='quantile')
with open(f"{OUT}/reliability.csv","w",newline="") as f:
    w=csv.writer(f); w.writerow(["method","pred","obs"])
    [w.writerow(["Original (slope 0.94)",a,b]) for a,b in zip(x0,o0)]
    [w.writerow(["Recalibrated (out-of-fold Platt)",a,b]) for a,b in zip(x1,o1)]
with open(f"{OUT}/cil.csv","w",newline="") as f:
    w=csv.writer(f); w.writerow(["label","cil","is_shd"])
    [w.writerow([NICE[j], y[:,j].mean()-probs[:,j].mean(), int(COLS[j]=='shd')]) for j in range(12)]
print("exported reliability.csv + cil.csv ->", OUT)
